### PR TITLE
Conjur JWT support

### PR DESCRIPTION
**Pull request by kbristow**
## Problem Statement

The Conjur Provider currently only has support for ApiKey authentication. This adds JWT authentication support as an additional option when using the Conjur Provider.

## Related Issue

Relates to #588. There is no specific issue open for adding JWT Auth support but can open one should it be required.

## Proposed Changes

This add the `jwt` option to the `auth` config for the SecretStore conjur provider. Users can now choose to use either `apikey` or `jwt` when using the Conjur integration. The PR includes:

- functionality to support using JWT auth with Conjur ESO Provider
- extension of the test suite to cover the new functionality
- updates to the documentation to explain usage of the jwt auth configuration

The implementation is heavily inspired from the vault and akeyless implementations for their jwt auth methods.

## Checklist

- [x] I have read the [contribution guidelines](https://external-secrets.io/latest/contributing/process/#submitting-a-pull-request)
- [x] All commits are signed with `git commit --signoff`
- [x] My changes have reasonable test coverage
- [x] All tests pass with `make test`
- [x] I ensured my PR is ready for review with `make reviewable`

### CODE REVIEW

**Comment by kbristow** [comment]
Maybe @davidh-cyberark you could also have a look at this or recommend someone from your side to have a look?

**Comment by davidh-cyberark** [comment]
I will start looking at this today.

**Review by davidh-cyberark — COMMENTED** [review]
overall, code looks ok.  I have not had a chance to do any testing, though.

Has there been any backward compatibility testing done to ensure apikey functionality still works as expected?  E.g. run tests in an actual environment and not just the unit-tests.

**Comment by kbristow** [comment]
> overall, code looks ok. I have not had a chance to do any testing, though.
> 
> Has there been any backward compatibility testing done to ensure apikey functionality still works as expected? E.g. run tests in an actual environment and not just the unit-tests.

I have been testing against our Conjur Enterprise instance which is running version 12.7. Everything appears to work without any regression except for a bug introduced on master unrelated to the this provder as I raised here: https://github.com/external-secrets/external-secrets/issues/2599. Happy if you would like to further testing than the scope I have available to me.

File: docs/provider/conjur.md [hunk 1]
```diff
@@ -9,32 +9,51 @@ This section contains the list of the pre-requirements before installing the Con
...
 
-### Create External Secret Store Definition
+### Certificate for Conjur server
 
-Recommend to save as filename: `conjur-secret-store.yaml`
+When using a self-signed cert when setting up your Conjur server, it is strongly recommended to populate "caBundle" with self-signed cert in the secret-store definition. The certificate CA must be referenced on the secret-store definition using either a `caBundle` or `caProvider` as below:
 
 ```yaml
-{% include 'conjur-secret-store.yaml' %}
+....
```
[the remark this answers is not in the document]
Comment (kbristow) [reply docs/provider/conjur.md:23]: Will take a look tomorrow or Monday.

**Comment by davidh-cyberark** [comment]
I plan to do more testing this week.  I should be able to post results by end of this week, Fri Aug 18.

**Comment by kbristow** [comment]
> I plan to do more testing this week. I should be able to post results by end of this week, Fri Aug 18.

Thanks for the feedback. I will keep an eye out for followups.

**Comment by kbristow** [comment]
Rebased onto main to include the fix for the orphaned secret logic mentioned earlier.

**Comment by kbristow** [comment]
> I plan to do more testing this week. I should be able to post results by end of this week, Fri Aug 18.

Hi @davidh-cyberark. Did you manage to take a look last week? Thanks.

**Comment by davidh-cyberark** [comment]
> > I plan to do more testing this week. I should be able to post results by end of this week, Fri Aug 18.
> 
> Hi @davidh-cyberark. Did you manage to take a look last week? Thanks.

Hi @kbristow, I did not get a chance to take a look last week due to other priorities.   

Testing is still on my todo list, however, I don't want to be a blocker, so, feel free to move forward if you think everything is ready.

**Comment by kbristow** [comment]
Thanks and no problem. I will start chasing reviews from ESO owners at the same time then.

@shuheiktgw - could you advise if you or someone from your side could take a look at this PR for me too please 🙏

**Comment by kbristow** [comment]
@shuheiktgw, @moolen - checking if you can spare sometime or recommend someone to give feedback.

**Comment by moolen** [comment]
sure, i'll take a look at it!

**Review by moolen — COMMENTED** [review]
Design+Code LGTM.

I consolidated the naming so it matches the other provider implementations (interface Provider/SecretsClient).
Left a comment regarding the expiry functionality. Would be great if we can remove it :crossed_fingers:

File: pkg/provider/conjur/provider.go [hunk 2]
```diff
@@ -36,64 +40,127 @@ var (
...
-	namespace    string
+	StoreKind        string
+	kube             client.Client
+	store            esv1beta1.GenericStore
+	namespace        string
+	corev1           typedcorev1.CoreV1Interface
+	clientAPI        ClientAPI
+	client           Client
+	clientExpires    bool
+	renewClientAfter time.Time
```
[the remark this answers is not in the document]
Comment (kbristow) [reply pkg/provider/conjur/provider.go:63]: Thanks. I was not aware that it was short lived. Happy to remove all that then. The life spans of the tokens would definitely be long enough for a reconcile of `Kind=ExternalSecret`.

[the remark this answers is not in the document]
Comment (kbristow) [reply pkg/provider/conjur/provider.go]: @moolen I have removed the renewal logic and the jwt parsing stuff. Let me know if you are happy with that.

**Comment by moolen** [comment]
Thanks @kbristow for the swift update :rocket:. I was trying to setup the `jwt` integration with `conjur`, but failed miserably :sweat:, can you provide me some guidance/examples/scripts or a walkthrough on how to set it up?

I was using [cyberark/conjur-quickstart](https://github.com/cyberark/conjur-quickstart) to get an integration environment, but failed to set it up for jwt integration. I followed [this guide](https://docs.conjur.org/Latest/en/Content/Integrations/k8s-ocp/k8s-jwt-authn.htm) but am failing to check the authenticator status. Any hints or alternative approaches appreciated :bow:

**Comment by kbristow** [comment]
> Thanks @kbristow for the swift update 🚀. I was trying to setup the `jwt` integration with `conjur`, but failed miserably 😓, can you provide me some guidance/examples/scripts or a walkthrough on how to set it up?
> 
> I was using [cyberark/conjur-quickstart](https://github.com/cyberark/conjur-quickstart) to get an integration environment, but failed to set it up for jwt integration. I followed [this guide](https://docs.conjur.org/Latest/en/Content/Integrations/k8s-ocp/k8s-jwt-authn.htm) but am failing to check the authenticator status. Any hints or alternative approaches appreciated 🙇

I have only tried the Enterprise version before, but they should be similar. I will take a look when I get a moment.

**Comment by davidh-cyberark** [comment]
> Had a look now, I unfortunately cannot run through the conjur quickstart since I have an m2 mac and it seems none of the conjur images are built for arm64. However, went and looked at the section of the guide you pointed to and the instructions there are quite sparse. To the check the status I think you need to first authenticate, then use the token you got from authenticating to check the status. Something like the following:
> 
> ```shell
> TOKEN=$(curl --header "Accept-Encoding: base64" --data <password-or-api-key> https://<conjur hostname>/authn/myorg/<username(probably admin)>/authenticate)
> curl -H "Authorization: Token token=\"${TOKEN}\"" \
>      https://<conjur hostname>/authn-jwt/<authenticator id>/myorg/status
> ```
> 
> @davidh-cyberark - would you guys be able to advise more for @moolen if he is still having trouble?

I've carved out time today to run tests against my dev conjur env.  I'll post results by tomorrow.

**Comment by moolen** [comment]
> @moolen is there anything more I can help to get this progressed?

sorry, i've attempted to set this up a couple of times, but due to the lack of knowledge and time i wasn't able to get this up and running. design+code looks good! If @davidh-cyberark can take a look that would be awesome :bow:. 

If not, we can merge it in if you tested it and are planning to use it in production :smile:

**Comment by kbristow** [comment]
> If not, we can merge it in if you tested it and are planning to use it in production 😄

We already use ESO in production :) just need to switch the secrets backend to Conjur.

**Comment by moolen** [comment]
going to merge this in now, this looks good to me!

**Review by moolen — APPROVED** [review]

**Pull request merged**